File: Solve-Path/Solve-Path/src/main/java/com/example/solvePath/DFS.java
package com.example.solvePath;

import java.util.Stack;
import java.util.Vector;

public class DFS {
    public Node startNode;
    public Node goalNode;
    public Node currentNode;
    private final Stack<Node> FRINGE;
    private Vector<Node> CHILD;
    public Vector<int[]> RESULT;
    protected int approvedNodes;
    protected int totalNodes;
    protected long time;
    protected static boolean stop = false;
    protected String error;

    public DFS() {
        FRINGE = new Stack<>();
        CHILD = new Vector<>();
        RESULT = new Vector<>();
    }

    public void solve(int maxDepth) {
        RESULT.clear();
        long startTime = System.currentTimeMillis();
        FRINGE.push(startNode);
        try {
            while (!FRINGE.isEmpty()) {
                currentNode = FRINGE.pop();

                // Kiểm tra thời gian
                if (System.currentTimeMillis() - startTime > 60000) {
                    error = "Thuật toán quá tốn thời gian!";
                    FRINGE.clear();
                    return;
                }

                // Kiểm tra trạng thái dừng
                if (stop) {
                    FRINGE.clear();
                    return;
                }

                // Kiểm tra nếu đã đạt đến node mục tiêu
                if (currentNode.equals(goalNode)) {
                    addResult(currentNode);
                    time = System.currentTimeMillis() - startTime;
                    totalNodes = approvedNodes + FRINGE.size();
                    FRINGE.clear();
                    return;
                }

                // Lấy các node con
                CHILD = currentNode.successors();

                // Lọc node cha khỏi danh sách node con
                CHILD.removeIf(child -> child.equals(currentNode.parent));

                // Thêm các node con vào FRINGE
                for (Node child : CHILD) {
                    child.parent = currentNode;
                    // Kiểm tra độ sâu trước khi thêm
                    if (child.cost <= maxDepth) {
                        FRINGE.push(child);
                    }
                }

                CHILD.clear();
                approvedNodes++;
            }
        } catch (OutOfMemoryError e) {
            FRINGE.clear();
            error = "Tràn bộ nhớ!";
            e.printStackTrace();
        }
    }

    // Truy vết kết quả
    public void addResult(Node n) {
        if (n.parent != null) {
            addResult(n.parent);
        }
        RESULT.add(n.state.value);
    }
}

File: NPuzzleAI/NPuzzleAI/DFS.cs
using System;
using System.Collections.Generic;

namespace NPuzzleAI
{
    public class DFS
    {
        public Node StartNode { get; set; }
        public Node GoalNode { get; set; }
        public Node CurrentNode { get; private set; }
        private readonly Stack<Node> fringe; // Thay Queue bằng Stack
        private List<Node> child; // Thay Vector bằng List
        public List<int[]> Result { get; private set; } // Thay Vector bằng List
        public int approvedNodes;
        public int totalNodes;
        public long time;
        public static bool stop = false;
        public string error;

        public DFS()
        {
            fringe = new Stack<Node>();
            child = new List<Node>();
            Result = new List<int[]>();
        }

        public void Solve()
        {
            Result.Clear();
            long startTime = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            fringe.Push(StartNode); // Sử dụng Push cho Stack
            try
            {
                while (fringe.Count > 0)
                {
                    CurrentNode = fringe.Pop(); // Sử dụng Pop cho Stack
                    if (DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - startTime > 60000)
                    {
                        error = "Thuật toán quá tốn thời gian!";
                        fringe.Clear();
                        return;
                    }
                    if (stop)
                    {
                        fringe.Clear();
                        return;
                    }
                    if (CurrentNode.Equals(GoalNode))
                    {
                        AddResult(CurrentNode);
                        time = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - startTime;
                        totalNodes = approvedNodes + fringe.Count;
                        fringe.Clear();
                        return;
                    }
                    child = CurrentNode.Successors();
                    if (CurrentNode.Parent != null)
                    {
                        for (int i = 0; i < child.Count; i++)
                        {
                            Node c = child[i];
                            if (c.Equals(CurrentNode.Parent))
                            {
                                child.RemoveAt(i);
                                break;
                            }
                        }
                    }
                    foreach (Node c in child)
                    {
                        c.Parent = CurrentNode;
                        fringe.Push(c); // Sử dụng Push để thêm vào Stack
                    }
                    child.Clear();
                    approvedNodes++;
                }
            }
            catch (OutOfMemoryException)
            {
                fringe.Clear();
                error = "Tràn bộ nhớ!";
            }
        }

        // Truy vết kết quả
        public void AddResult(Node n)
        {
            if (n.Parent != null)
            {
                AddResult(n.Parent);
            }
            Result.Add(n.State.Value);
        }
    }
}
